### PR TITLE
Add support for function references (free and static member) in bind_back

### PR DESCRIPTION
**Pull request by denizevrenci**
The following program does not compile: https://godbolt.org/z/WfhEc51dY But if we replace the function reference with a lambda, it does: https://godbolt.org/z/jrhTY4nse Constructing a std::tuple with a function type fails but function pointers are allowed. We instantiate _result with std::decay_t<ArgN>... which maps function references to function pointers and then the corresponding std::tuple instantiation does not fail to compile. The problem with the existing code is that, in order to instantiate std::tuple<std::decay_t<ArgN>...>, we first constructed a std::tuple<ArgN> and passed it over. If instead we construct a std::tuple<std::decay_t<ArgN>...> with the function arguments, they will be converted to function pointers and the std::tuple instantiation will compile.

### CODE REVIEW

File: test/let_done_test.cpp [hunk 1]
```diff
@@ -93,3 +93,52 @@ TEST(TransformDone, WithValue) {
...
 }
+
+static auto just42() {
+  return just(42);
+}
+
+TEST(TransformDone, WithFunction) {
+  auto freeFunction = just_done() | let_done(just42) | sync_wait();
+
+  EXPECT_TRUE(freeFunction.has_value());
```
Comment (ispeters) [inline test/let_done_test.cpp:104]: I like to use `ASSERT_TRUE` for this assertion; if the assertion is ever false, `EXPECT_TRUE` will report the error and continue execution, leading to UB on the next line but `ASSERT_TRUE` will report the error and then end the test (by executing a `return;` statement).
```suggestion
  ASSERT_TRUE(freeFunction.has_value());
```

File: test/let_done_test.cpp [hunk 2]
```diff
@@ -93,3 +93,52 @@ TEST(TransformDone, WithValue) {
...
+  EXPECT_EQ(*freeFunction, 42);
+
+  struct StaticMemberFunction {
+    static auto call() { return just(42, 1, 2); }
+  };
+
+  auto staticMemberFunction =
+      just_done() | let_done(StaticMemberFunction::call) | sync_wait();
+
+  EXPECT_TRUE(staticMemberFunction.has_value());
```
Comment (ispeters) [inline test/let_done_test.cpp:114]:
```suggestion
  ASSERT_TRUE(staticMemberFunction.has_value());
```

File: test/let_done_test.cpp [hunk 3]
```diff
@@ -93,3 +93,52 @@ TEST(TransformDone, WithValue) {
...
+    ExplicitCopy() = default;
+
+    explicit ExplicitCopy(const ExplicitCopy& other) = default;
+
+    auto operator()() { return just(42, 1, 2); }
+  };
+
+  auto explicitCopy = just_done() | let_done(ExplicitCopy{}) | sync_wait();
+
+  EXPECT_TRUE(explicitCopy.has_value());
```
Comment (ispeters) [inline test/let_done_test.cpp:129]:
```suggestion
  ASSERT_TRUE(explicitCopy.has_value());
```

File: test/let_done_test.cpp [hunk 4]
```diff
@@ -93,3 +93,52 @@ TEST(TransformDone, WithValue) {
...
+    ExplicitMove() = default;
+
+    explicit ExplicitMove(ExplicitMove&& other) = default;
+
+    auto operator()() { return just(42, 1, 2); }
+  };
+
+  auto explicitMove = just_done() | let_done(ExplicitMove{}) | sync_wait();
+
+  EXPECT_TRUE(explicitMove.has_value());
```
Comment (ispeters) [inline test/let_done_test.cpp:142]:
```suggestion
  ASSERT_TRUE(explicitMove.has_value());
```

File: include/unifex/detail/concept_macros.hpp [hunk 5]
```diff
@@ -387,6 +387,13 @@ UNIFEX_CONCEPT  //
     callable =  //
     sizeof(decltype(_is_callable::_try_call(static_cast<Fn (*)(As...)>(
         nullptr)))) == sizeof(_is_callable::yes_type);
+
+template <typename Ret, typename... As>
+UNIFEX_CONCEPT              //
+    callable<Ret(As...)> =  //
+    sizeof(decltype(_is_callable::_try_call(
+        static_cast<std::add_pointer_t<Ret(As...)> (*)(As...)>(nullptr)))) ==
+    sizeof(_is_callable::yes_type);
```
Comment (ispeters) [inline include/unifex/detail/concept_macros.hpp:396]: What's this for? I _think_ the explicit specialization is misaligned with the intended usage.

**Review by ispeters — COMMENTED** [review]
A couple of nits regarding the tests, but I think we should chat about the new `callable` specialization.

File: include/unifex/detail/concept_macros.hpp [hunk 6]
```diff
@@ -387,6 +387,13 @@ UNIFEX_CONCEPT  //
     callable =  //
     sizeof(decltype(_is_callable::_try_call(static_cast<Fn (*)(As...)>(
```
Comment (ispeters) [inline include/unifex/detail/concept_macros.hpp:388]:
```suggestion
    sizeof(decltype(_is_callable::_try_call(static_cast<std::decay_t<Fn> (*)(As...)>(
```

File: include/unifex/at_coroutine_exit.hpp [hunk 7]
```diff
@@ -313,7 +313,7 @@ inline constexpr struct _fn {
 
 public:
   template(typename Action, typename... Ts)                           //
-      (requires callable<std::decay_t<Action>, std::decay_t<Ts>...>)  //
+      (requires invocable<std::decay_t<Action>, std::decay_t<Ts>...>)  //
```
Comment (ispeters) [inline include/unifex/at_coroutine_exit.hpp:316]: Should this be `std::invocable_v`?
```suggestion
      (requires std::invocable_v<std::decay_t<Action>, std::decay_t<Ts>...>)  //
```

Comment (denizevrenci) [reply include/unifex/at_coroutine_exit.hpp:316]: It's using our concept shim https://github.com/facebookexperimental/libunifex/blob/d5639a2bb88b8a5375a9c4e8665e0111f2ca336a/include/unifex/detail/concept_macros.hpp#L377-L385

File: include/unifex/v1/async_scope.hpp [hunk 8]
```diff
@@ -398,7 +398,7 @@ struct async_scope {
    * Equivalent to attach_on((Scheduler&&)scheduler, just_from((Fun&&)fun)).
    */
   template(typename Scheduler, typename Fun)             //
-      (requires scheduler<Scheduler> AND callable<Fun>)  //
+      (requires scheduler<Scheduler> AND invocable<Fun>)  //
```
Comment (ispeters) [inline include/unifex/v1/async_scope.hpp:401]:
```suggestion
      (requires scheduler<Scheduler> AND std::is_invocable_v<Fun>)  //
```

File: include/unifex/v1/debug_async_scope.hpp [hunk 9]
```diff
@@ -44,7 +44,7 @@ struct debug_async_scope final {
   }
 
   template(typename Scheduler, typename Fun)             //
-      (requires scheduler<Scheduler> AND callable<Fun>)  //
+      (requires scheduler<Scheduler> AND invocable<Fun>)  //
```
Comment (ispeters) [inline include/unifex/v1/debug_async_scope.hpp:47]:
```suggestion
      (requires scheduler<Scheduler> AND std::is_invocable_v<Fun>)  //
```

File: include/unifex/v1/debug_async_scope.hpp [hunk 10]
```diff
@@ -64,11 +64,11 @@ struct debug_async_scope final {
...
```
Comment (ispeters) [inline include/unifex/v1/debug_async_scope.hpp:67]:
```suggestion
      (requires scheduler<Scheduler> AND std::is_invocable_v<Fun>)  //
```

File: include/unifex/v1/debug_async_scope.hpp [hunk 11]
```diff
@@ -90,7 +90,7 @@ struct debug_async_scope final {
   }
 
   template(typename Fun)        //
-      (requires callable<Fun>)  //
+      (requires invocable<Fun>)  //
```
Comment (ispeters) [inline include/unifex/v1/debug_async_scope.hpp:93]:
```suggestion
      (requires std::is_invocable_v<Fun>)  //
```

File: include/unifex/v1/debug_async_scope.hpp [hunk 12]
```diff
@@ -107,7 +107,7 @@ struct debug_async_scope final {
   }
 
   template(typename Scheduler, typename Fun)             //
-      (requires scheduler<Scheduler> AND callable<Fun>)  //
+      (requires scheduler<Scheduler> AND invocable<Fun>)  //
```
Comment (ispeters) [inline include/unifex/v1/debug_async_scope.hpp:110]:
```suggestion
      (requires scheduler<Scheduler> AND std::is_invocable_v<Fun>)  //
```

File: include/unifex/with_query_value.hpp [hunk 13]
```diff
@@ -48,13 +48,13 @@ class _receiver_wrapper<CPO, Value, Receiver>::type {
   }
 
   template(typename OtherCPO, typename Self, typename... Args)  //
-      (requires same_as<remove_cvref_t<Self>, type> AND callable<
+      (requires same_as<remove_cvref_t<Self>, type> AND invocable<
```
Comment (ispeters) [inline include/unifex/with_query_value.hpp:51]:
```suggestion
      (requires same_as<remove_cvref_t<Self>, type> AND std::is_invocable_v<
```

File: test/any_object_test.cpp [hunk 14]
```diff
@@ -340,7 +340,7 @@ static_assert(std::is_convertible_v<foo_supported, any_foo>);
 static_assert(!std::is_convertible_v<int, any_foo>);
 
 // Some other checks about the nature of the type-erased wrapper.
-static_assert(unifex::callable<get_foo_cpo, const any_foo&>);
+static_assert(unifex::invocable<get_foo_cpo, const any_foo&>);
```
Comment (ispeters) [inline test/any_object_test.cpp:343]:
```suggestion
static_assert(std::is_invocable_v<get_foo_cpo, const any_foo&>);
```

**Review by ispeters — COMMENTED** [review]
Looking good!

File: include/unifex/at_coroutine_exit.hpp (hunk 7 above)
Comment (denizevrenci) [reply include/unifex/at_coroutine_exit.hpp:316]: Okay. My updates were me trying to get that to work. But I'll just drop it altogether then.

File: include/unifex/win32/windows_thread_pool.hpp [hunk 15]
```diff
@@ -703,7 +703,7 @@ class windows_thread_pool::_schedule_at_op<Receiver>::type final
 
 private:
   void set_value_impl() noexcept override {
-    if constexpr (unifex::is_nothrow_callable_v<
+    if constexpr (unifex::std::is_nothrow_invocable_v<
```
Comment (ispeters) [inline include/unifex/win32/windows_thread_pool.hpp:706]:
```suggestion
    if constexpr (std::is_nothrow_invocable_v<
```

File: include/unifex/win32/windows_thread_pool.hpp [hunk 16]
```diff
@@ -791,7 +791,7 @@ class windows_thread_pool::_schedule_after_op<Duration, Receiver>::type final
...
```
Comment (ispeters) [inline include/unifex/win32/windows_thread_pool.hpp:794]:
```suggestion
    if constexpr (std::is_nothrow_invocable_v<
```

File: include/unifex/spawn_future.hpp [hunk 17]
```diff
@@ -699,7 +699,7 @@ struct _future_sender_from_stop_token final {
   struct type;
 };
 
-// A callable that returns a Sender given an inplace_stop_token.  An instance of
+// A invocable that returns a Sender given an inplace_stop_token.  An instance of
```
[the remark this answers is not in the document]
Comment (denizevrenci) [reply include/unifex/spawn_future.hpp:702]: Search and replace gone wrong.

**Review by ispeters — APPROVED** [review]
LGTM

**Pull request merged**